Toggle side-by-side rendering for current notebook

echarles opened the pull request:
## References

https://github.com/jupyterlab/jupyterlab/issues/11376

## Code changes

Add a command, a shortcut to toggle the side-by-side rendering only for the current notebook.

## User-facing changes

![Untitled](https://user-images.githubusercontent.com/226720/144471373-11e64491-6df6-44ea-a85b-49df8151c2c8.gif)

## Backwards-incompatible changes

None

fcollonval reviewed (commented):
LGTM

One question, would it make sense to be easier to switch on a per notebook basis and more complex to do it globally?
Say otherwise should this new shortcut fall to `Shift R` and the older to `Ctrl Shift R`?

echarles commented:
> Say otherwise should this new shortcut fall to Shift R and the older to Ctrl Shift R?

Makes a lot sense to me. May also this should be added to the menu, but I fear some kind of `menu congestion`... I will wait a bit more for more feedbacks on these.

<img width="317" alt="Screenshot 2021-12-03 at 10 05 20" src="https://user-images.githubusercontent.com/226720/144575837-0f41d546-105f-4cbe-9375-54e1fff9e646.png">

fcollonval commented:
We probably want to keep only the action for the active notebook in the menu as this is already the case for items like _show line numbers_, _match brackets_,...

echarles commented:
@fcollonval I have pushed the needed change to toogle for the current notebook with `Shift R`.

A few questions.

## 1. Notebook metadata

The notebook reminds the side-by-side state via metadata. For now, after save, a new metadata is persisted. It is not clear to me from https://nbformat.readthedocs.io/en/latest/format_description.html if this is allowed/accepted. I don't know either if we want that. I can see advantages to it, as the next notebook author designs its content to fit a defined presentation, so the next reader get benefit from that persisted stat

<img width="310" alt="Screenshot 2021-12-06 at 08 10 57" src="https://user-images.githubusercontent.com/226720/144806541-4c99282a-2d53-42cd-9506-d05c3c235c0e.png">
e

## 2. RTC

I have not tested with RTC, but I can imagine your notebook rendering suddenly changing if a remote collaborator toggles its side-by-side state. Do we want to prevent that or it is a welcome feature? 

## 3. Menu label

For now, the label in the menu is a bit long, though descriptive `Toggle Side-by-side rendering for current Notebook`. Any comment on that? 

<img width="459" alt="Screenshot 2021-12-06 at 08 11 56" src="https://user-images.githubusercontent.com/226720/144806799-41703386-0e78-4579-b1be-51fad637cee3.png">

fcollonval commented:
> @fcollonval I have pushed the needed change to toogle for the current notebook with `Shift R`.

Thanks

> ## 1. Notebook metadata
> 
> The notebook reminds the side-by-side state via metadata. For now, after save, a new metadata is persisted. It is not clear to me from https://nbformat.readthedocs.io/en/latest/format_description.html if this is allowed/accepted. I don't know either if we want that. I can see advantages to it, as the next notebook author designs its content to fit a defined presentation, so the next reader get benefit from that persisted stat

I don't know. But it used to be done like this by lots of extensions in the classical notebook.

I realize also that the convention goes more towards using `_` rather than `-`.


> ## 2. RTC
> 
> I have not tested with RTC, but I can imagine your notebook rendering suddenly changing if a remote collaborator toggles its side-by-side state. Do we want to prevent that or it is a welcome feature?

As far as I understand RTC, only a subset of explicitly given metadata are synchronized between clients. cc @hbcarlos 

> ## 3. Menu label
> 
> For now, the label in the menu is a bit long, though descriptive `Toggle Side-by-side rendering for current Notebook`. Any comment on that?

What about dropping the `for current Notebook` and use `Toggle Side-by-Side Rendering` (using capitalized word for homogeneity)? The command will not be enable on other document, isn't?

hbcarlos commented:
@fcollonval thanks for pinging me.

Since #11064 all the metadata is shared except for `kernelspec`, `language_info`.

fcollonval commented:
Thanks Carlos

Then indeed synchronizing those view related settings may be pretty annoying as they are more a question of taste.

echarles commented:
@fcollonval @hbcarlos thx for the reviews and comments. I have pushed those changes

1. Notebook metadata: Nows uses `render_side_by_side` (with underscores)
2. RTC: Added `render_side_by_side` to the `UNSHARED_KEYS` to prevent unexpected UI changes
3. Label: Uses `Toggle Side-by-side Rendering` (short and correctly defines the action)

fcollonval commented:
Thanks @echarles - for me you only need to update the UI test snapshot and we are good to go.

krassowski commented:
> 2. RTC: Added `render_side_by_side` to the `UNSHARED_KEYS` to prevent unexpected UI changes

- Which one gets ultimately saved? The one of the host? If collaborator reopens they will see different layout, right?
- What if I want to have two views of the same notebook, do they have to be both in the same mode (side-by-side or normal)?

I think we should discuss whether this should be per-view or per-notebook setting; for majority of use cases they will behave the same in a single session, except for the RTC corner cases and multiple views of the same notebook). The drawback of having it per view/widget is that we cannot save the state (well we can save it in layout restorer but it won't be persistent beyond current workspace).

I can see an argument against storing it along: we do not store scroll position in notebook itself (although it could be useful) because it is the responsibility of the viewer to remember the state (in our case of the layout restorer).

There is also an analogy with other documents (such as Word): do they store the view state in the documents themselves (e.g. web view, print view, normal view; zoom level)? I think that some of these things are stored, but I am not sure if all.

krassowski commented:
Instead of `render_side_by_side` we should **consider** using something more generic like `rendering_layout: "side_by_side"` or `arrange: "side_by_side"` (is there a preference for underscore/hyphen in values too?), so that we can add more mutually exclusive modes in the future without having to worry about what to do when both `render_side_by_side` and say `render_detached` (or whichever we will explore next) mean when both are set.

Some users like to run notebooks one cell at a time; I could imagine a view in which the active cell is always centered and output is displayed in dedicated output widget/panel as in RStudio. Or just a view identical as in RStudio (so input cells are mostly as they are but output is always shown in a separate tab which can be moved around). But then those views are more like user-preference layouts (as discussed in #8960) and we probably would not want to save them in the notebook anyways (?).

Another question is whether we should distinguish left-to-right and right-to-left (i.e. whether outputs are on the left or on the right) positioning; this could be an extra metadata field to be added in the future, or just different values like `layout: "side-by-side-outputs-on-left"`. This is important from authoring perspective (if users write "see the result on the right but someone will be viewing this in a localized LTR view in the future they might be confused).

echarles commented:
@krassowski 

1. If we don't use the notebook metadata to store the `render_side_by_side` rendering, I guess we will need to store it at the Panel/Widget level. Do you see other options.
2. We can move to a more `rendering_layout` naming.

echarles commented:
@krassowski @fcollonval @hbcarlos I have pushed more to generalize the rendering layout notion and not depend on the notebook metadata anymore. This means you can toggle now the notebook views independently.

![Untitled](https://user-images.githubusercontent.com/226720/145168876-fd565467-2bba-467b-b48e-7d82f1796c84.gif)

echarles commented:
@krassowski I hope I have addressed your points with my last commit. The menu is not changed (still using the `Toogle side-by-side rendering`) as this is the only rendering mode for now, but the implementation supports adding more rendering modes.

fcollonval commented:
I played with this PR on binder and found three issues (one not directly related to this PR but to side-by-side feature):

![bug_side_by_side](https://user-images.githubusercontent.com/8435071/145541410-ac8ad808-1d90-4b07-bd95-c74bb7570b8f.gif)

1. The screencast displays a notebook with a second view on the same notebook. If I hit `Shift R` on the left notebook, then focus on the right notebook; the right notebook switch to side-by-side rendering without pressing the shortcut... This is unexpected. It should either switch directly or not at all.
2. Long text in code-mirror do not reduce their width as other cells (look at the second code cell).
3. If the screen width is small (like in the screencast) the sides are overlapping. I would forbid using side by side mode if the width is too small (although _too small_ is a complex notion to translate in numbers...).

echarles commented:
Thx @fcollonval I see indeed non-coheren behavior when hitting `Shift-R`. I don't see a strict reproducible pattern, and have not hit that during the dev of that feature. I will look at that.

Your points 2. and 3. are IMHO not related to this feature.

echarles commented:
> Thx @fcollonval I see indeed non-coherent behavior when hitting Shift-R. I don't see a strict reproducible pattern, and have not hit that during the dev of that feature. I will look at that.

@fcollonval Latest commit fixes point 1. We can open separated issues for points 2. and 3. (we need to discuss how the global settings act in relationship with the toggling for such small screen cases)

fcollonval commented:
> We can open separated issues for points 2. and 3. (we need to discuss how the global settings act in relationship with the toggling for such small screen cases)

I'm fine with that being tackle in another PR (just wanted to point it out).

fcollonval commented:
The CI error seems relevant

```
Error: test/widget.spec.ts(54,3): error TS2741: Property 'renderingLayout' is missing in type '{ showHiddenCellsButton: boolean; scrollPastEnd: boolean; defaultCell: nbformat.CellType; recordTiming: boolean; numberCellsToRenderDirectly: number; renderCellOnIdle: boolean; observedTopMargin: string; observedBottomMargin: string; maxNumberOutputs: number; disableDocumentWideUndoRedo: boolean; }' but required in type 'INotebookConfig'.
```

echarles commented:
> I'm fine with that being tackle in another PR (just wanted to point it out).

Just opened 2 follow-up issues for that:

- https://github.com/jupyterlab/jupyterlab/issues/11652
- https://github.com/jupyterlab/jupyterlab/issues/11653

echarles commented:
Test is fixed. Last CI run shows 3 failing which I think are not introduced by this PR.

fcollonval reviewed (approved):
LGTM

> UI test failure is not related.

echarles commented:
Thx @fcollonval for the reviews and approval.

@krassowski anything missing to get this merged?

jtpio commented:
@echarles I restarted the failing checks just to be sure

echarles commented:
@jtpio Result is `4 cancelled and 62 successful checks`

krassowski commented on packages/notebook-extension/src/index.ts:
```diff
@@ -2347,29 +2353,36 @@ function addCommands(
 
   commands.addCommand(CommandIDs.toggleRenderSideBySide, {
-    label: trans.__('Render Side-by-side'),
+    label: trans.__('Toggle Side-by-side rendering setting'),
```
```suggestion
    label: trans.__('Render Side-by-side by Default'),
```

Or some variation? Like I don't feel like it's much better, but the current version is not distinct enough from the per-notebook command IMO.

echarles replied:
The `toggleRenderSideBySide` is responsible to toggle the setting. I am not sure that `Render Side-by-side by Default` expresses that action, especially that the sentence remains the same whatever the setting state (true/false) is.

krassowski replied:
Yes, my proposal is not great. But I do feel like "toggle setting" will not be obvious to the user. Maybe something like "Toggle default notebook layout: side-by-side"?

echarles replied:
I have no strong opinion on that. Maybe input from @fcollonval or @jtpio would help.

jtpio replied:
Another option could be `Enable Side-by-side Rendering by Default`?

echarles replied:
I have removed that command which had been there in the previous iteration of the side-by-side, but looking closer, was not changing for real the user setting, so was not doing the job it was supposed to do.

Also we have generalized the the notion of rendering layout, so it become difficult anyway to find a good fit for that commande.

fcollonval replied:
Earlier I proposed to use `Render Side-by-Side` (using the :white_check_mark: mark to distinguish the status) to be aligned with for example `Show line numbers` (that is also per document) or `Wrap words`.

echarles replied:
Hi, thx for the help here. To wrap up, we have a setting `renderingLayout` which can be `default` or `side-by-side` and we are talking about 2 commands

- The one that changes toggle the layout of the current notebook and is available via the menu. The name is now `Toggle Side-by-side Rendering`. I can change it to `Render Side-by-Side` with the ✅.
- The one that was there before and I have removed which was supposed to toggle the user setting. The former implementation had just a in-memory flag that was used to apply the toggle the rendering on all the notebooks, but was note changing the user setting. I don't think a command can, or should change a user setting. This is why I have dropped it

jtpio commented on packages/notebook/test/widget.spec.ts:
```diff
@@ -43,7 +43,8 @@ const notebookConfig = {
   maxNumberOutputs: 50,
-  disableDocumentWideUndoRedo: true
+  disableDocumentWideUndoRedo: true,
+  renderingLayout: 'default' as 'default' | 'side-by-side'
```
Maybe `'default' | 'side-by-side'` should be made a `type` for improved readability and maintainability?

echarles replied:
Done. Thx @jtpio

echarles commented:
My last commit implements what we have been discussed on https://github.com/jupyterlab/jupyterlab/pull/11608#discussion_r770284417

The menu shows a mark to indicate the side-by-side toggling state 

![Untitled](https://user-images.githubusercontent.com/226720/146349306-7293180f-04e5-47bc-9c6f-af501c46fc3c.gif)

I think we are good to go now.

fcollonval commented:
I'm :+1: with the latest version. The reference snapshots for the UI tests need to be updated. Then this is good to me.

echarles commented:
I have updated the UI snapshot. UI Tests are failing with the following (before it was explicitly failing with the wrong snapshots. I will relaunch to see what it gives.

```
    test-results/test-jupyterlab-debugger-Debugger-Tests-Start-debug-session-jupyterlab/cc0cedab382e97654b8e1332220a4967.webm
    ------------------------------------------------------------------------------------------------

    Retry #1 ---------------------------------------------------------------------------------------

    page.evaluate: TypeError: Cannot read properties of null (reading 'type')

        at GalataInpage.waitForCellRun (<anonymous>:3790:37)
```

echarles commented:
RTD is still building with Error Concurrency limit reached (2), retrying in 5 minutes.

Otherwise all green.

echarles commented:
@fcollonval CI is now fully green. Can we merge this?

echarles commented:
Thx @fcollonval @krassowski @jtpio for the help.

echarles commented:
@meeseeksdev please backport to 3.2.x
